envoy/buffer/buffer.h

```diff
@@ -458,6 +458,40 @@ class Instance {
     writeInt<ByteOrder::BigEndian, T, Size>(value);
   }
 
+  /**
+   * Copy multiple arguments to the buffer. The type of arguments must be automatically convertible
+   * (e.g. char*, char[], std::string) to `absl::string_view`.
+   * @param args A set of arguments with variable length and type. The arguments must be guaranteed
+   * to be safely convertible to `absl::string_view`.
+   * @return The total size of the data copied to the buffer.
+   */
+  template <bool ActionAfterAdd = true, class... Args> size_t addFragments(Args&&... args) {
+    size_t total_size_to_write = 0;
+
+    static_assert(sizeof...(args) >= 2, "At least two arguments");
+
+    constexpr auto write_helper = [](uint8_t** dst, absl::string_view v) {
+      memcpy(*dst, v.data(), v.size()); // NOLINT(safe-memcpy)
+      *dst += v.size();
+    };
+
+    total_size_to_write = (absl::string_view(args).size() + ...);
+
+    auto* mem = inlineReserve(total_size_to_write);
+    if (mem != nullptr) {
+      (write_helper(&mem, absl::string_view(args)), ...);
+      // Call the `add` method to trigger action after writing such as watermark checking.
+      if constexpr (ActionAfterAdd) {
+        add(nullptr, 0);
+      }
+    } else {
+      // No enough continuous memory in the back slice and downgrade to using `add` method.
+      (add(absl::string_view(args)), ...);
```

Comment:
Mea culpa -- I have never used this C++ syntax before in my decades with the language :) So I might be misunderstanding this fragment, but it looks like this might be a perf gap.

Let me contrive an example. Let's say you only have 1 byte remaining in the back slice, and we are adding 1000 1-byte fragments in one call. It looks to me like we'll do 1000 slow `add()`s in that case, whereas what we'd really like is to fill the remaining bytes in the back slice, and then continue to do addFragments after that, until the last fragment when we'll do the final `add()`.

I might be misunderstanding the syntax though.

Also I don't necessarily expect this PR to resolve all possible perf scenarios, but if my analysis makes sense, maybe you could at least put as TODO in to attack that case in a follow-up?

Reply:
You are quite correct. However, the number of parameters should not normally exceed ten, so it should not be the case that you are talking about. 🤔 
If we want to support a large number of fragments (say hundreds) being written at the same time, perhaps we need an interface that accepts containers (std::vector<absl::string_view> or absl::Span<absl::string_view> or something).
I think this might also be and is valuable to work with.

Reply:
> I think my suggestion here, in the context of the current API you defined, is that if the first fragment spills over the slice, we don't watermark-check for every subsequent arg. I don't know how often that might happen; it might be so rare we should ignore it, but it's hard to tell looking at the code.

This is related to the size of fragments. If the length of a single fragment is relatively small, it will not overflow frequently. If the length of the fragment is large, then it is still worth an add()`.

But I also agree that there is still some room for optimization here. But I haven't figured out how to do it yet.